python/cvmfs/repository.py

```diff
@@ -197,39 +196,230 @@ def _pop_catalog(self):
         return wrapper.get_catalog()
 
 
+class Resource(file):
+    def __init__(self, name, mode, tmp_dir):
+        self.__final_destination_path = name
+        temp_file_path = tempfile.mktemp(dir=tmp_dir, prefix='tmp.')
+        super(Resource, self).__init__(temp_file_path, mode)
+
+    def close(self):
+        super(Resource, self).close()
+        os.rename(self.name, self.__final_destination_path)
+
+
+class Cache(object):
+    def __init__(self, cache_dir):
+        if not os.path.exists(cache_dir):
+            cache_dir = tempfile.mkdtemp(dir='/tmp', prefix='cache.')
+        self._cache_dir = cache_dir
+        self._create_cache_structure()
+        self._cleanup_metadata()
+
+    def _cleanup_metadata(self):
+        metadata_file_list = [
+            os.path.join(self._cache_dir, _common._MANIFEST_NAME),
+            os.path.join(self._cache_dir, _common._LAST_REPLICATION_NAME),
+            os.path.join(self._cache_dir, _common._REPLICATING_NAME),
+            os.path.join(self._cache_dir, _common._WHITELIST_NAME)
+        ]
+        for metadata_file in metadata_file_list:
+            try:
+                os.remove(metadata_file)
+            except OSError:
+                pass
+
+    def _create_dir(self, path):
+        cache_full_path = os.path.join(self._cache_dir, path)
+        if not os.path.exists(cache_full_path):
+            os.mkdir(cache_full_path, 0755)
+
+    def _create_cache_structure(self):
+        self._create_dir('data')
+        for i in range(0x00, 0xff + 1):
+            new_folder = '{0:#0{1}x}'.format(i, 4)[2:]
+            self._create_dir(os.path.join('data', new_folder))
+        self._create_dir(os.path.join('data', 'txn'))
+
+    def get_transaction_dir(self):
+        return os.path.join(self._cache_dir, 'data', 'txn')
+
+    def get_cache_path(self):
+        return str(self._cache_dir)
+
+    def add(self, file_name):
+        full_path = os.path.join(self._cache_dir, file_name)
+        tmp_dir = self.get_transaction_dir()
+        return Resource(full_path, "w+", tmp_dir)
+
+    def get(self, file_name):
+        full_path = os.path.join(self._cache_dir, file_name)
+        if os.path.exists(full_path):
+            return open(full_path, 'rb')
+        return None
+
+    def clear(self):
+        data_path = os.path.join(self._cache_dir, 'data')
+        shutil.rmtree(data_path)
+        self._create_cache_structure()
+
+
+class Fetcher(object):
+    """ Abstract wrapper around a Fetcher """
+
+    __metadata__ = abc.ABCMeta
+
+    def __init__(self, source, cache_dir=''):
+        self.__cache = Cache(cache_dir)
+        self.source = source
+
+    def _make_file_uri(self, file_name):
+        return os.path.join(self.source, file_name)
+
+    def get_cache_path(self):
+        return self.__cache.get_cache_path()
+
+    def retrieve_file(self, file_name):
```

Comment:
How about that to avoid the code duplication in `retrieve_file()` and `retrieve_raw_file()` and to make the cache control flow more obvious:

```python
def retrieve_file(self, file_name):
    return self._retrieve(file_name, self._retrieve_file)

def retrieve_raw_file(self, file_name):
    return self._retrieve(file_name, self._retrieve_raw_file)

def retrieve_file(self, file_name, retrieve_fn):
    cached_file = self.__cache.get(file_name)
    if not cached_file:
        cached_file = self.__cache.transaction(file_name)
        retrieve_fn(file_name, cached_file)
        cached_file = self.__cache.commit(file_name)
    return cached_file
```

You probably still need the `Resource` class to keep the final destination path but perhaps you can tightly couple it to `Cache` by making it a member class of it.